File: web/src/handTracker.js
/* handTracker.js — MediaPipe Hands wrapper with EMA smoothing and adaptive thresholds */

const dist2D = (p1, p2) => Math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2);

// Exponential Moving Average helper
function ema(current, previous, alpha) {
  return alpha * current + (1 - alpha) * previous;
}

export class HandTracker {
  constructor(frameW = 1280, frameH = 720) {
    this.W = frameW;
    this.H = frameH;

    // ── Smoothing ────────────────────────────────────────────────────────────
    // Lower = smoother but more lag. 0.25 is a good balance for game control.
    this.SMOOTH_ALPHA = 0.25;

    // ── Pinch Detection ──────────────────────────────────────────────────────
    // Normalized (0-1) threshold — adapts automatically to hand distance.
    // ~0.06 units in normalized coords ≈ fingers just touching regardless of depth.
    this.PINCH_THRESHOLD_NORM = 0.07;

    // Require this many consecutive frames to confirm a pinch (debounce)
    this.PINCH_DEBOUNCE_FRAMES = 3;
    this.pinchConsecutive = 0;
    this.releaseConsecutive = 0;
    this.pinchActive = false;    // debounced pinch state

    // ── Z-Push Detection ─────────────────────────────────────────────────────
    this.Z_CLICK_THRESHOLD_M = 0.012;
    this.Z_CLICK_XY_MAX_PX   = 35;
    this.zHistoryLen          = 10;
    this.zCooldownFrames      = 25;

    // State
    this.zHistory  = [];
    this.zCooldown = 0;
    this.zStartXY  = null;

    // ── Smoothed positions (subpixel) ────────────────────────────────────────
    this._smoothIX = -1;  // -1 = uninitialised
    this._smoothIY = -1;
    this._smoothPX = -1;
    this._smoothPY = -1;

    // ── ToF Sensor State ─────────────────────────────────────────────────────
    this.tofActive    = false;
    this.tofSimulated = false;

    // ── Finger Lock State ───────────────────────────────────────────────────
    this._lockedPos  = null;
    this._lostFrames = 0;

    // ── Output ────────────────────────────────────────────────────────────────
    this.gesture = this._emptyGesture();
    this.hands   = null;
  }

  toggleTofSimulation() {
    this.tofSimulated = !this.tofSimulated;
    return this.tofSimulated;
  }

  sampleTofDepth(zVal) {
    if (this.tofActive || this.tofSimulated) {
      const zM = Number(Math.max(0.15, 0.45 + zVal * 0.6).toFixed(3));
      const label = this.tofActive ? "ToF IR Hardware" : "ToF Hardware (Simulated)";
      return { active: true, depthM: zM, source: label };
    }
    return { active: false, depthM: 0, source: "RGB MediaPipe Estimate" };
  }

  _emptyGesture() {
    return {
      hand_visible:      false,
      index_pos:         [0, 0],
      pinch_pos:         [0, 0],
      is_pinching:       false,
      click_just_fired:  false,
      is_index_isolated: false,
      zDelta:            0,
      xyDrift:           0,
      tof_active:        false,
      tof_z_m:           0,
      depth_source:      "RGB MediaPipe Estimate",
    };
  }

  /**
   * Initialize MediaPipe Hands.
   * @param {Function} onReadyCallback
   */
  init(onReadyCallback) {
    const MP_Hands = window.Hands;
    if (!MP_Hands) {
      console.error('[ERROR] MediaPipe Hands not loaded from CDN.');
      return;
    }

    this.hands = new MP_Hands({
      locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`,
    });

    this.hands.setOptions({
      maxNumHands:            2,
      modelComplexity:        1,
      minDetectionConfidence: 0.75,
      minTrackingConfidence:  0.75,   // raised from 0.6 → fewer jittery re-acquisitions
    });

    this.hands.onResults((results) => this._processResults(results));

    this.hands.initialize()
      .then(() => {
        console.log('[INFO] MediaPipe Hands initialized.');
        if (onReadyCallback) onReadyCallback();
      })
      .catch((err) => {
        console.error('[ERROR] MediaPipe Hands init failed:', err);
      });
  }

  async process(videoElement) {
    if (this.hands) await this.hands.send({ image: videoElement });
  }

  // ── Internal result processor ─────────────────────────────────────────────

  _processResults(results) {
    const g = this._emptyGesture();

    if (results.multiHandLandmarks && results.multiHandLandmarks.length > 0) {
      const lm = this._selectLockedHand(results.multiHandLandmarks);
      if (lm) {

      // ── Raw normalized coords (no flooring — keep subpixel) ───────────────
      const normX = (id) => (1 - lm[id].x) * this.W;   // mirror horizontally
      const normY = (id) =>      lm[id].y  * this.H;

      const rawIX = normX(8);  // index tip
      const rawIY = normY(8);
      const rawTX = normX(4);  // thumb tip
      const rawTY = normY(4);

      // ── EMA smoothing on landmark positions ───────────────────────────────
      // Bootstrap on first frame
      if (this._smoothIX < 0) {
        this._smoothIX = rawIX;
        this._smoothIY = rawIY;
        this._smoothPX = (rawIX + rawTX) / 2;
        this._smoothPY = (rawIY + rawTY) / 2;
      }

      // ── 2. Z-Push / Click Detection ───────────────────────────────────────
      const zVal = lm[8].z;
      const { fired, delta, drift } = this._detectZClick(zVal, [Math.round(this._smoothIX), Math.round(this._smoothIY)]);
      g.click_just_fired = fired;
      g.zDelta           = delta;
      g.xyDrift          = drift;

      // Heavy dampening during forward Z-push to prevent lateral cursor drift
      const currentAlpha = delta > 0.015 ? 0.02 : this.SMOOTH_ALPHA;

      this._smoothIX = ema(rawIX, this._smoothIX, currentAlpha);
      this._smoothIY = ema(rawIY, this._smoothIY, currentAlpha);

      const rawPX = (rawIX + rawTX) / 2;
      const rawPY = (rawIY + rawTY) / 2;
      this._smoothPX = ema(rawPX, this._smoothPX, currentAlpha);
      this._smoothPY = ema(rawPY, this._smoothPY, currentAlpha);

      // Output as rounded integers for pixel consumption, but smoothed first
      g.hand_visible = true;
      g.index_pos    = [Math.round(this._smoothIX), Math.round(this._smoothIY)];
      g.pinch_pos    = [Math.round(this._smoothPX), Math.round(this._smoothPY)];

      // ── 1. Pinch Detection (normalized + debounced) ───────────────────────
      const pinchDistNorm = Math.sqrt(
        (lm[4].x - lm[8].x) ** 2 +
        (lm[4].y - lm[8].y) ** 2 +
        (lm[4].z - lm[8].z) ** 2 * 0.5   // z has lower weight (noisier)
      );

      const rawPinching = pinchDistNorm < this.PINCH_THRESHOLD_NORM;

      // Debounce: require N consecutive frames before toggling state
      if (rawPinching) {
        this.pinchConsecutive++;
        this.releaseConsecutive = 0;
        if (this.pinchConsecutive >= this.PINCH_DEBOUNCE_FRAMES) {
          this.pinchActive = true;
        }
      } else {
        this.releaseConsecutive++;
        this.pinchConsecutive = 0;
        if (this.releaseConsecutive >= this.PINCH_DEBOUNCE_FRAMES) {
          this.pinchActive = false;
        }
      }

      g.is_pinching = this.pinchActive;

      // ── 3. Index Finger Isolation (curl-based) ────────────────────────────
      const dist3D = (a, b) => Math.sqrt(
        (lm[a].x - lm[b].x) ** 2 +
        (lm[a].y - lm[b].y) ** 2 +
        (lm[a].z - lm[b].z) ** 2
      );

      // Palm reference: landmark 9 (middle finger MCP) — stable anchor
      const isExtended = (tip, pip, mcp) =>
        dist3D(tip, 9) > dist3D(mcp, 9) * 0.85;

      const indexExt  = isExtended(8,  6,  5);
      const middleExt = isExtended(12, 10, 9);
      const ringExt   = isExtended(16, 14, 13);
      const pinkyExt  = isExtended(20, 18, 17);

      // Relaxed index isolation: allow middle finger co-extension (natural tendon attachment)
      g.is_index_isolated = indexExt && !(ringExt || pinkyExt);

      // ToF Depth sampling
      const tofRes = this.sampleTofDepth(lm[8].z);
      g.tof_active   = tofRes.active;
      g.tof_z_m      = tofRes.depthM;
      g.depth_source = tofRes.source;

      } else {
        this._lostFrames++;
        if (this._lostFrames > 12) {
          this._smoothIX = -1;
          this._smoothIY = -1;
          this._lockedPos = null;
        }
      }
    } else {
      // Hand lost — reset all smoothing and detection state
      this._lostFrames++;
      if (this._lostFrames > 12) {
        this._smoothIX      = -1;
        this._smoothIY      = -1;
        this._smoothPX      = -1;
        this._smoothPY      = -1;
        this.pinchConsecutive   = 0;
        this.releaseConsecutive = 0;
        this.pinchActive    = false;
        this.zHistory       = [];
        this.zStartXY       = null;
        this._lockedPos     = null;
      }
    }

    this.gesture = g;
  }

  _selectLockedHand(multiHandLandmarks) {
    const MAX_LOCK_DIST_PX = 250.0;
    const candidates = multiHandLandmarks.map((lm) => {
      const ix = (1 - lm[8].x) * this.W;
      const iy = lm[8].y * this.H;
      return { ix, iy, lm };
    });

    if (candidates.length === 0) return null;

    if (!this._lockedPos) {
      this._lockedPos = [candidates[0].ix, candidates[0].iy];
      this._lostFrames = 0;
      return candidates[0].lm;
    }

    const [lx, ly] = this._lockedPos;
    let bestLm = null;
    let bestDist = Infinity;
    let bestPos = null;

    for (const c of candidates) {
      const dist = Math.hypot(c.ix - lx, c.iy - ly);
      if (dist < bestDist) {
        bestDist = dist;
        bestLm = c.lm;
        bestPos = [c.ix, c.iy];
      }
    }

    if (bestDist <= MAX_LOCK_DIST_PX) {
      this._lockedPos = bestPos;
      this._lostFrames = 0;
      return bestLm;
    }

    return null;
  }

  // ── Z-Push click algorithm ────────────────────────────────────────────────

  _detectZClick(zNow, xyNow) {
    const result = { fired: false, delta: 0, drift: 0 };

    if (this.zCooldown > 0) {
      this.zCooldown--;
      this.zHistory.push(zNow);
      if (this.zHistory.length > this.zHistoryLen) this.zHistory.shift();
      return result;
    }

    this.zHistory.push(zNow);
    if (this.zHistory.length > this.zHistoryLen) this.zHistory.shift();
    if (this.zHistory.length < this.zHistoryLen)  return result;

    const zBaseline = this.zHistory[0];
    const deltaZ    = zBaseline - zNow;  // positive = finger moving toward camera
    result.delta    = deltaZ;

    if (this.zStartXY !== null) {
      result.drift = dist2D(this.zStartXY, xyNow);
    }

    if (deltaZ >= this.Z_CLICK_THRESHOLD_M) {
      if (this.zStartXY === null) {
        this.zStartXY = xyNow;
        result.drift  = 0;
      } else {
        result.drift  = dist2D(this.zStartXY, xyNow);
      }

      if (result.drift < this.Z_CLICK_XY_MAX_PX) {
        // Confirmed push-click
        this.zHistory  = [];
        this.zStartXY  = null;
        this.zCooldown = this.zCooldownFrames;
        result.fired   = true;
      } else {
        // Too much lateral drift — not a clean push
        this.zHistory = [];
        this.zStartXY = null;
      }
    } else {
      this.zStartXY = xyNow;
    }

    return result;
  }
}

File: birds-game/hand_tracker.py
"""
hand_tracker.py — MediaPipe hands wrapper.
Gesture detection:
  1. Pinch click  : thumb-tip ↔ index-tip distance < PINCH_THRESHOLD px
  2. Z-push click : index fingertip moves ~1 inch toward the camera (−Z)
                    while staying nearly still in X/Y (< Z_CLICK_XY_MAX_PX)
Both gestures produce the same "click" event so game.py doesn't need to care
which method was used.
"""
import math
import cv2
import mediapipe as mp
import numpy as np
from physics import (
    PINCH_THRESHOLD,
    Z_CLICK_THRESHOLD_M,
    Z_CLICK_XY_MAX_PX,
    distance,
)

# Number of history frames we keep for Z-push detection
_Z_HISTORY = 8          # ~8 frames at 30 fps ≈ 0.27 s window
_Z_COOLDOWN_FRAMES = 20  # frames before another z-click can fire


class HandTracker:
    """
    Wraps MediaPipe Hands.  Call process(frame) each frame, read .gesture dict.
    gesture keys
    hand_visible  : bool
    index_pos     : (px, py) – index fingertip in pixel space
    pinch_pos     : (px, py) – midpoint of thumb & index tip
    is_pinching   : bool  – thumb/index close together
    click_just_fired : bool – True for exactly ONE frame when either click fires
    """

    def __init__(self, frame_w=1280, frame_h=720):
        self.W = frame_w
        self.H = frame_h
        self._mp_hands = mp.solutions.hands
        self._hands = self._mp_hands.Hands(
            max_num_hands=2,
            min_detection_confidence=0.7,
            min_tracking_confidence=0.6,
        )
        # Z history for push-click  (raw MediaPipe Z values, negative = closer)
        self._z_history: list[float] = []
        self._z_cooldown = 0  # countdown in frames
        # XY snapshot at the moment Z started changing (for drift check)
        self._z_start_xy: tuple[float, float] | None = None
        self._z_start_val: float | None = None

        # Finger lock state
        self._locked_pos: tuple[float, float] | None = None
        self._lost_frames: int = 0

        # Exposed result
        self.gesture: dict = self._empty_gesture()
        self.tof_simulated: bool = False
        self.tof_active: bool    = False

    # ── public ────────────────────────────────────────────────────────────────
    def toggle_tof_simulation(self) -> bool:
        """Toggle simulated ToF depth stream on/off for debugging."""
        self.tof_simulated = not self.tof_simulated
        return self.tof_simulated

    def sample_tof_depth(self, z_val: float) -> tuple[bool, float, str]:
        """Look up true physical depth Z at pinch location."""
        if self.tof_active or self.tof_simulated:
            z_m = round(max(0.15, 0.45 + (z_val * 0.6)), 3)
            label = "ToF IR Hardware" if self.tof_active else "ToF Hardware (Simulated)"
            return True, z_m, label
        return False, 0.0, "RGB MediaPipe Estimate"

    def process(self, bgr_frame: np.ndarray) -> dict:
        """Run detection on one BGR frame (already flipped).  Returns gesture."""
        rgb = cv2.cvtColor(bgr_frame, cv2.COLOR_BGR2RGB)
        result = self._hands.process(rgb)

        g = self._empty_gesture()

        if result.multi_hand_landmarks:
            selected_lm = self._select_locked_hand(result.multi_hand_landmarks)
            if selected_lm is not None:
                lm = selected_lm.landmark

            # ── landmark positions ─────────────────────────────────────────
            def px(lm_id):
                return int(lm[lm_id].x * self.W), int(lm[lm_id].y * self.H)
            thumb_tip  = px(4)
            index_tip  = px(8)
            midpoint   = (
                (thumb_tip[0] + index_tip[0]) // 2,
                (thumb_tip[1] + index_tip[1]) // 2,
            )
            g["hand_visible"] = True
            g["index_pos"]    = index_tip
            g["pinch_pos"]    = midpoint

            # ── 1. Pinch click ────────────────────────────────────────────
            pinch_dist = distance(thumb_tip, index_tip)
            g["is_pinching"] = pinch_dist < PINCH_THRESHOLD

            # ── 2. Z-push click ───────────────────────────────────────────
            z_val = lm[8].z          # MediaPipe Z: negative = closer to camera
            click_fired, z_delta, xy_drift = self._detect_z_click(z_val, index_tip)
            g["click_just_fired"] = click_fired
            g["z_delta"]          = z_delta
            g["xy_drift"]         = xy_drift

            # ── ToF depth lookup at pinch point ───────────────────────────
            tof_act, tof_z, tof_src = self.sample_tof_depth(z_val)
            g["tof_active"]   = tof_act
            g["tof_z_m"]      = tof_z
            g["depth_source"] = tof_src

            # ── 3. Index isolation ────────────────────────────────────────
            def dist_sq_from_wrist(tip_id, pip_id):
                # extended if tip is further from wrist (0) than pip
                d_tip = (lm[tip_id].x - lm[0].x)**2 + (lm[tip_id].y - lm[0].y)**2
                d_pip = (lm[pip_id].x - lm[0].x)**2 + (lm[pip_id].y - lm[0].y)**2
                return d_tip > d_pip
                
            index_ext = dist_sq_from_wrist(8, 6)
            middle_ext = dist_sq_from_wrist(12, 10)
            ring_ext = dist_sq_from_wrist(16, 14)
            pinky_ext = dist_sq_from_wrist(20, 18)
            
            # Relaxed index isolation: allow middle finger co-extension (natural tendon attachment)
            g["is_index_isolated"] = index_ext and not (ring_ext or pinky_ext)

            self._lost_frames = 0

        else:
            # No hand / lost lock: increment lost frames and reset if grace period expires
            self._lost_frames += 1
            if self._lost_frames > 12:
                self._z_history.clear()
                self._z_start_xy  = None
                self._z_start_val = None
                self._locked_pos   = None

        self.gesture = g
        return g
    def _select_locked_hand(self, multi_hand_landmarks):
        """
        Locks onto a single index finger. If multiple hands are detected,
        returns the landmark set whose index tip is closest to the locked position.
        """
        MAX_LOCK_DIST_PX = 250.0

        candidates = []
        for hand_lms in multi_hand_landmarks:
            lm = hand_lms.landmark
            ix = lm[8].x * self.W
            iy = lm[8].y * self.H
            candidates.append((ix, iy, hand_lms))

        if not candidates:
            return None

        if self._locked_pos is None:
            best_ix, best_iy, best_lms = candidates[0]
            self._locked_pos = (best_ix, best_iy)
            return best_lms

        lx, ly = self._locked_pos
        best_lms = None
        best_dist = float('inf')
        best_pos = None

        for ix, iy, hand_lms in candidates:
            dist = math.sqrt((ix - lx) ** 2 + (iy - ly) ** 2)
            if dist < best_dist:
                best_dist = dist
                best_lms = hand_lms
                best_pos = (ix, iy)

        if best_dist <= MAX_LOCK_DIST_PX:
            self._locked_pos = best_pos
            return best_lms

        return None

    # ── private ───────────────────────────────────────────────────────────────
    @staticmethod
    def _empty_gesture() -> dict:
        return {
            "hand_visible":      False,
            "index_pos":         (0, 0),
            "pinch_pos":         (0, 0),
            "is_pinching":       False,
            "click_just_fired":  False,
            "is_index_isolated": False,
            "z_delta":           0.0,
            "xy_drift":          0.0,
            "tof_active":        False,
            "tof_z_m":           0.0,
            "depth_source":      "RGB MediaPipe Estimate",
        }

    def _detect_z_click(self, z_now: float, xy_now: tuple[int, int]) -> tuple[bool, float, float]:
        """
        Returns (fired: bool, delta_z: float, xy_drift: float).
        """
        drift = 0.0

        if self._z_cooldown > 0:
            self._z_cooldown -= 1
            self._z_history.append(z_now)
            if len(self._z_history) > _Z_HISTORY:
                self._z_history.pop(0)
            return False, 0.0, 0.0

        self._z_history.append(z_now)
        if len(self._z_history) > _Z_HISTORY:
            self._z_history.pop(0)

        if len(self._z_history) < _Z_HISTORY:
            return False, 0.0, 0.0

        z_start = self._z_history[0]   # oldest sample in window
        delta_z = z_start - z_now      # positive = moved toward camera

        if self._z_start_xy is not None:
            drift = distance(self._z_start_xy, xy_now)

        if delta_z >= Z_CLICK_THRESHOLD_M:
            if self._z_start_xy is None:
                self._z_start_xy = xy_now
                drift = 0.0
            else:
                drift = distance(self._z_start_xy, xy_now)

            if drift < Z_CLICK_XY_MAX_PX:
                # Valid Z-push click!
                self._z_history.clear()
                self._z_start_xy  = None
                self._z_start_val = None
                self._z_cooldown  = _Z_COOLDOWN_FRAMES
                return True, delta_z, drift
            else:
                # Too much XY movement → treat as hand movement, reset
                self._z_history.clear()
                self._z_start_xy  = None
        else:
            # Push not started yet — update XY baseline each frame
            self._z_start_xy = xy_now

        return False, delta_z, drift

    def draw_landmarks(self, frame: np.ndarray):
        """Optional debug overlay — draws hand skeleton."""
        mp.solutions.drawing_utils.draw_landmarks(
            frame,
            self._mp_hands.HandLandmark if False else None,  # skip
        )
